src/Processors/Formats/Impl/AvroRowInputFormat.cpp

```diff
@@ -438,6 +434,39 @@ AvroDeserializer::DeserializeFn AvroDeserializer::createDeserializeFn(const avro
                         target_type->getName());
                 }
             }
+
+            if (target.isVariant())
+            {
+                const DataTypeVariant & variant_type = assert_cast<const DataTypeVariant &>(*target_type);
+                const auto & nested_types = variant_type.getVariants();
+
+                if (root_node->leaves() != nested_types.size())
+                    throw Exception(ErrorCodes::INCORRECT_DATA, "The number of union types in Avro record ({}) doesn't match the number of types in destination Variant type ({}).", root_node->leaves(), nested_types.size());
+
+                std::vector<DeserializeFn> nested_deserializers;
+                nested_deserializers.reserve(root_node->leaves());
+
+                for (int i = 0; i != static_cast<int>(root_node->leaves()); ++i)
+                {
+                    auto nested_deserializer = createDeserializeFn(root_node->leafAt(i), nested_types.at(i));
```

Comment:
No need to copy or move the closure here. Please, replace with the inplace construct nested_deserializers.emplace_back(createDeserializeFn(root_node->leafAt(i), nested_types.at(i)))

Reply:
Moved.